homework/index.js

```diff
@@ -29,18 +33,275 @@
+    });
+    createAndAppend('i', divWidgetRepoUpdatedData, {
+      class: 'widget-icon fas fa-clock fa-3x px-3 py-1',
+    });
+    createAndAppend('h4', divWidgetRepoUpdatedData, {
+      text: 'Updated:',
+      class: 'py-2 px-2 widget-title',
+    });
+    const updatedDate = new Date(repositoriesData.updated_at);
+    createAndAppend('h4', divWidgetRepoUpdatedData, {
+      text: updatedDate.toDateString(),
+      class: 'widget-value lead',
+    });
+    //
+    const divWidgetForkedNumbers = createAndAppend('div', leftColumn, {
+      class:
+        'd-flex flex-column  alert alert-primary justify-content-start align-items-center flex-sm-column flex-lg-row',
+    });
+    createAndAppend('i', divWidgetForkedNumbers, {
+      class: 'widget-icon fas fa-code-branch fa-3x px-3 py-1',
+    });
+    createAndAppend('h4', divWidgetForkedNumbers, {
+      text: 'Forks:',
+      class: 'py-2 px-2 widget-title',
+    });
+    createAndAppend('h4', divWidgetForkedNumbers, {
+      text: repositoriesData.forks,
+      class: 'widget-value lead',
+    });
+  };
+  // This function is created to reduce callback hell length.
+  const contributionStats = (parentElement, cardBody, dataArr, textValue) => {
+    let parent = parentElement;
+    parent = createAndAppend('p', cardBody, {
+      class: 'card-text btn btn-danger d-flex justify-content-between rounded-0',
+      text: `${textValue}:`,
+    });
+    createAndAppend('span', parent, {
+      class: 'card-text badge badge-light d-flex justify-content-between ',
+      text: dataArr.length,
+    });
+  };
+  //
+  const appendRepositoriesToSelect = (repositoriesData, cb) => {
+    const selectList = document.getElementById('repo-select');
+    repositoriesData
+      .sort((a, b) => a.name.localeCompare(b.name))
+      .forEach((repository, index) => {
+        createAndAppend('option', selectList, {
+          text: repository.name,
+          value: index,
+        });
+      });
 
-  function main(url) {
-    fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
+    // create an eventlistener for select list options
+
+    selectList.addEventListener('change', () => {
+      selectList.parentElement.nextElementSibling.innerHTML = '';
+      cb(repositoriesData[selectList.value], repositoriesData[selectList.value].contributors_url);
+    });
+  };
+  // fetch repositories and contributions data as a callback hell so that we are sure once repositories data fetched first, then the contribution data fetched.
+  // if it is not used callback hell, the contribution stats such as followers, following and subscriptions couldnt be created on the same order thats why we need callback hell
+  // if i knew promises i think the code would be nicer :)
+  const fetchAllData = url => {
```

Comment:
You are right that promises will enable you to write your code more elegantly. You can still get rid of the deep nesting by using a style called **continuation-passing** (once you master promises no need for this anymore). Basically, in this style you create separate named function at the global/file level for all your (async) callbacks. If those functions require access to inner variables for the calling function you can pass them through a technique called 'currying' (google for it): basically a parameterized function returning a function. Here is your reworked `fetchData()` function using this technique (the names for callbacks might be improved).

> Note that the GitHub API limits you to 60 requests per hour (more is possible but you need to use an API key). Your code can easily hit that limit if you run it a couple of times.

```js
const followingsCallback = (cardBody, subscriptions) => (errFollowing, followingsArr) => {
  if (errFollowing) {
    createAndAppend('div', root, {
      text: errFollowing.message,
      class: 'alert alert-danger display-3 text-center',
    });
    return;
  }
  contributionStats('followingInfo', cardBody, followingsArr, 'Following');
  fetchJSON(subscriptions, (errSubscriptions, subscriptionsArr) => {
    if (errSubscriptions) {
      createAndAppend('div', root, {
        text: errSubscriptions.message,
        class: 'alert alert-danger display-3 text-center',
      });
      return;
    }
    contributionStats('subscriptionInfo', cardBody, subscriptionsArr, 'Subscriptions');
  });
};

const followersCallback = (subscriptions, followings, cardBody) => (
  errFollowers,
  followersArr,
) => {
  if (errFollowers) {
    createAndAppend('div', root, {
      text: errFollowers.message,
      class: 'alert alert-danger display-3 text-center',
    });
    return;
  }
  contributionStats('followersInfo', cardBody, followersArr, 'Followers');
  fetchJSON(followings, followingsCallback(cardBody, subscriptions));
};

const contributionsCallback = (errContributions, contributionsObj) => {
  if (errContributions) {
    createAndAppend('div', root, {
      text: `${errContributions.message}... The data cannot be fetched`,
      class: 'alert alert-danger display-3 text-center',
    });
    return;
  }
  createContributionsLayout(
    contributionsObj,
    (followers, followings, subscriptions, cardBody) => {
      fetchJSON(followers, followersCallback(subscriptions, followings, cardBody));
    },
  );
};

const repositoriesCallback = (errRepository, repositoriesData) => {
  if (errRepository) {
    createAndAppend('div', root, {
      text: errRepository.message,
      class: 'alert alert-danger display-3 text-center',
    });
    return;
  }
  appendRepositoriesToSelect(repositoriesData, (repositoryObj, contributionsURL) => {
    createRepositoryLayout(repositoryObj);
    fetchJSON(contributionsURL, contributionsCallback);
  });
};

const fetchAllData = url => {
  fetchJSON(url, repositoriesCallback);
};
```